man/gemstash-configuration.5.md
---
title: gemstash-configuration
date: October 13, 2015
section: 5
...

# Name

gemstash-configuration

# Synopsis

```yaml
# ~/.gemstash/config.yml
---
:base_path: "/var/gemstash"
:cache_type: memcached
:memcached_servers: localhost:11211
:db_adapter: postgres
:db_url: postgres:///gemstash
:db_connection_options:
  :test: true
  :pool_timeout: 2
:rubygems_url: https://my.gem-source.local
:ignore_gemfile_source: false
:puma_threads: 32
:bind: tcp://0.0.0.0:4242
:protected_fetch: true
:fetch_timeout: 10
:open_timeout: 2
:log_file: gemstash.log
```

# Base Path

`:base_path`

Specifies where to store local files like the server log, cached gem files, and
the database (when using SQLite). If the default is being used, the directory
will be created if it does not exist. Any other directory needs to be created
ahead of time and be writable to the Gemstash server process. Specifying the
`:base_path` via [`gemstash setup`][SETUP] will create the directory for you.

## Default value

`~/.gemstash`

## Valid values

Any valid path

# Cache Type

`:cache_type`

Specifies how to cache values other than gem files (such as gem dependencies).
`memory` will use an in memory cache while `memcached` will point to 1 or more
Memcached servers. Use the `:memcached_servers` configuration key for specifying
where the Memcached server(s) are.

## Default value

`memory`

## Valid values

`memory`, `memcached`

# Memcached Servers

`:memcached_servers`

Specifies the Memcached servers to connect to when using `memcached` for the
`:cache_type`. Only used when `memcached` is used for `:cache_type`.

## Default value

`localhost:11211`

## Valid values

A comma delimited list of Memcached servers

# DB Adapter

`:db_adapter`

Specifies what database adapter to use. When `sqlite3` is used, the database
will be located at `gemstash.db` within the directory specified by `:base_path`.
The database will automatically be created when using `sqlite3`. When
`postgres`, `mysql`, or `mysql2` is used, the database to connect to must be
specified in the `:db_url` configuration key. The database must already be
created when using anything other than `sqlite3`.

## Default value

`sqlite3`

## Valid values

`sqlite3`, `postgres`, `mysql`, `mysql2`

# DB URL

`:db_url`

Specifies the database to connect to when using `postgres`, `mysql`, or `mysql2`
for the `:db_adapter`. Only used when the `:db_adapter` is not `sqlite3`.

## Default value

None

## Valid values

A valid database URL for the [Sequel gem][SEQUEL]

# DB Connection Options

`:db_connection_options`

Specifies additional `Sequel.connect` options to use. Note that any options here
are merged in with the default options, so you need not specify the
`max_connections` if you customize this option.

## Default value

`{ max_connections: 1 }` for `sqlite3` adapter,
`{ max_connections: config[:puma_threads] + 1 }` for any other adapter.

## Valid values

A valid connection options Hash for the [Sequel.connect][SEQUEL_CONNECT] method.

# Rubygems URL

`:rubygems_url`

Specifies the default gem source URL. When any API endpoint is called without a
`/private` or `/upstream/<url>` prefix, this URL will be used to fetch the
result. This value can be safely changed even if there are already gems stashed
for the previous value.

## Default value

`https://rubygems.org`

## Valid values

A valid gem source URL

# Ignore Gemfile source

`:ignore_gemfile_source`

Ignore the source specified in Gemfile and always use `:rubygems_url` as gems upstream.

## Default value

`false`

## Valid values

Boolean: `true` or `false`

# Puma Threads

`:puma_threads`

Specifies the number of threads used for the Gemstash server.

## Default value

`16`

## Valid values

Integer value with a minimum of `1`

# Bind Address

`:bind`

Specifies the binding used to start the Gemstash server. Keep in mind the user
starting Gemstash needs to have access to bind in this manner. For example, if
you use a port below 1024, you will need to run Gemstash as the root user.

## Default value

`tcp://0.0.0.0:9292`

## Valid values

Any valid binding that [is supported by Puma][PUMA_BINDING]

# Protected Fetch

`:protected_fetch`

Tells Gemstash to authenticate via an API key before allowing the fetching of
private gems and specs. The default behavior is to allow unauthenticated
download of private gems and specs.

## Default value

`false`

## Valid values

Boolean values `true` or `false`

# Fetch Timeout

`:fetch_timeout`

This is the number of seconds to allow for fetching a gem from upstream.
It covers establishing the connection and receiving the response. Fetching
gems over a slow connection may cause timeout errors. If you experience
timeout errors, you may want to increase this value. The default is `20`
seconds.

## Default value

`20`

## Valid values

Integer value with a minimum of `1`

# Open Timeout

`:open_timeout`

The timeout setting for opening the connection to an upstream gem
server. On high-latency networks, even establishing the connection
to an upstream gem server can take a while. If you experience
connection failures instead of timeout errors, you may want to
increase this value. The default is `2` seconds.

## Default value

`2`

## Valid values

Integer value with a minimum of `1`

# Log File

`:log_file`

Indicates the name of the file to use for logging. The file will be placed in
the [base path][BASE_PATH].

## Default value

`server.log`

## Valid values

Any valid file name, or `:stdout` to log to `$stdout`

[SETUP]: ./gemstash-setup.1.md
[SEQUEL]: http://sequel.jeremyevans.net/
[SEQUEL_CONNECT]: http://sequel.jeremyevans.net/rdoc/files/doc/opening_databases_rdoc.html#label-General+connection+options
[PUMA_BINDING]: https://github.com/puma/puma#binding-tcp--sockets
[BASE_PATH]: ./gemstash-configuration.5.md#base-path


docs/gemstash-configuration.5.md
<!-- Automatically generated by Pandoc -->


# Name

gemstash-configuration

# Synopsis

``` yaml
# ~/.gemstash/config.yml
---
:base_path: "/var/gemstash"
:cache_type: memcached
:memcached_servers: localhost:11211
:db_adapter: postgres
:db_url: postgres:///gemstash
:db_connection_options:
  :test: true
  :pool_timeout: 2
:rubygems_url: https://my.gem-source.local
:ignore_gemfile_source: false
:puma_threads: 32
:bind: tcp://0.0.0.0:4242
:protected_fetch: true
:fetch_timeout: 10
:open_timeout: 2
:log_file: gemstash.log
```

# Base Path

`:base_path`

Specifies where to store local files like the server log, cached gem
files, and the database (when using SQLite). If the default is being
used, the directory will be created if it does not exist. Any other
directory needs to be created ahead of time and be writable to the
Gemstash server process. Specifying the `:base_path` via
[`gemstash setup`](docs/gemstash-setup.1.md) will create the directory for
you.

## Default value

`~/.gemstash`

## Valid values

Any valid path

# Cache Type

`:cache_type`

Specifies how to cache values other than gem files (such as gem
dependencies). `memory` will use an in memory cache while `memcached`
will point to 1 or more Memcached servers. Use the `:memcached_servers`
configuration key for specifying where the Memcached server(s) are.

## Default value

`memory`

## Valid values

`memory`, `memcached`

# Memcached Servers

`:memcached_servers`

Specifies the Memcached servers to connect to when using `memcached` for
the `:cache_type`. Only used when `memcached` is used for `:cache_type`.

## Default value

`localhost:11211`

## Valid values

A comma delimited list of Memcached servers

# DB Adapter

`:db_adapter`

Specifies what database adapter to use. When `sqlite3` is used, the
database will be located at `gemstash.db` within the directory specified
by `:base_path`. The database will automatically be created when using
`sqlite3`. When `postgres`, `mysql`, or `mysql2` is used, the database
to connect to must be specified in the `:db_url` configuration key. The
database must already be created when using anything other than
`sqlite3`.

## Default value

`sqlite3`

## Valid values

`sqlite3`, `postgres`, `mysql`, `mysql2`

# DB URL

`:db_url`

Specifies the database to connect to when using `postgres`, `mysql`, or
`mysql2` for the `:db_adapter`. Only used when the `:db_adapter` is not
`sqlite3`.

## Default value

None

## Valid values

A valid database URL for the [Sequel
gem](http://sequel.jeremyevans.net/)

# DB Connection Options

`:db_connection_options`

Specifies additional `Sequel.connect` options to use. Note that any
options here are merged in with the default options, so you need not
specify the `max_connections` if you customize this option.

## Default value

`{ max_connections: 1 }` for `sqlite3` adapter,
`{ max_connections: config[:puma_threads] + 1 }` for any other adapter.

## Valid values

A valid connection options Hash for the
[Sequel.connect](http://sequel.jeremyevans.net/rdoc/files/doc/opening_databases_rdoc.html#label-General+connection+options)
method.

# Rubygems URL

`:rubygems_url`

Specifies the default gem source URL. When any API endpoint is called
without a `/private` or `/upstream/<url>` prefix, this URL will be used
to fetch the result. This value can be safely changed even if there are
already gems stashed for the previous value.

## Default value

`https://rubygems.org`

## Valid values

A valid gem source URL

# Ignore Gemfile source

`:ignore_gemfile_source`

Ignore the source specified in Gemfile and always use `:rubygems_url` as
gems upstream.

## Default value

`false`

## Valid values

Boolean: `true` or `false`

# Puma Threads

`:puma_threads`

Specifies the number of threads used for the Gemstash server.

## Default value

`16`

## Valid values

Integer value with a minimum of `1`

# Bind Address

`:bind`

Specifies the binding used to start the Gemstash server. Keep in mind
the user starting Gemstash needs to have access to bind in this manner.
For example, if you use a port below 1024, you will need to run Gemstash
as the root user.

## Default value

`tcp://0.0.0.0:9292`

## Valid values

Any valid binding that [is supported by
Puma](https://github.com/puma/puma#binding-tcp--sockets)

# Protected Fetch

`:protected_fetch`

Tells Gemstash to authenticate via an API key before allowing the
fetching of private gems and specs. The default behavior is to allow
unauthenticated download of private gems and specs.

## Default value

`false`

## Valid values

Boolean values `true` or `false`

# Fetch Timeout

`:fetch_timeout`

This is the number of seconds to allow for fetching a gem from upstream.
It covers establishing the connection and receiving the response. Fetching
gems over a slow connection may cause timeout errors. If you experience
timeout errors, you may want to increase this value. The default is `20`
seconds.

## Default value

`20`

## Valid values

Integer value with a minimum of `1`

# Open Timeout

`:open_timeout`

The timeout setting for opening the connection to an upstream gem
server. On high-latency networks, even establishing the connection
to an upstream gem server can take a while. If you experience
connection failures instead of timeout errors, you may want to
increase this value. The default is `2` seconds.

## Default value

`2`

## Valid values

Integer value with a minimum of `1`

# Log File

`:log_file`

Indicates the name of the file to use for logging. The file will be
placed in the [base path](docs/gemstash-configuration.5.md#base-path).

## Default value

`server.log`

## Valid values

Any valid file name, or `:stdout` to log to `$stdout`


CHANGELOG.md
## 2.7.1 (2023-10-12)

### Fixes

  - Reinstate NOOP --daemonize and --no-daemonize that warn about them doing nothing. ([#381](https://github.com/rubygems/gemstash/pull/381) [@olleolleolle](https://github.com/olleolleolle))

## 2.7.0 (2023-10-08)

### Fixes

  - Fix uninitialized constant error when using Redis cache ([#375](https://github.com/rubygems/gemstash/pull/375), [@chris72205](https://github.com/chris72205))
  - Remove --daemonize option, no longer supported by Puma. ([#359](https://github.com/rubygems/gemstash/issues/359) [@olleolleolle](https://github.com/olleolleolle))

## 2.6.0 (2023-09-30)

### Changes

  - Support `GEMSTASH_CONFIG` environment variable ([#369](https://github.com/rubygems/gemstash/pull/369), [@kyrofa](https://github.com/kyrofa))

## 2.5.0 (2023-09-28)

### Changes

  - Add support for upstream auth with ENV variables ([#339](https://github.com/rubygems/gemstash/pull/339), [@CiTroNaK](https://github.com/CiTroNaK))

## 2.4.0 (2023-09-27)

### Changes

  - Support naming and listing authorizations ([#365](https://github.com/rubygems/gemstash/pull/365), [@kyrofa](https://github.com/kyrofa))

## 2.3.2 (2023-09-14)

### Fixes

  - Require a now-needed file for Puma 6. Thanks, @ktreis! ([#362](https://github.com/rubygems/gemstash/pull/362), [@olleolleolle](https://github.com/olleolleolle))

## 2.3.1 (2023-09-05)

### Fixes

  - Reinstate Ruby Central in README. ([#353](https://github.com/rubygems/gemstash/pull/353), [@olleolleolle](https://github.com/olleolleolle))
  - Implement Puma 5-required logger method `#sync` ([#355](https://github.com/rubygems/gemstash/pull/355), [@ball-hayden](https://github.com/ball-hayden))

## 2.3.0 (2023-08-30)

### Changes

  - Support Puma 6. ([#335](https://github.com/rubygems/gemstash/pull/335), [@MSP-Greg](https://github.com/MSP-Greg))
  - Make UpstreamSource intermediate specs from upstream. ([#337](https://github.com/rubygems/gemstash/pull/337), [@tedgarb](https://github.com/tedgarb))
  - Mention RubyCentral. ([#338](https://github.com/rubygems/gemstash/pull/338), [@adarsh](https://github.com/adarsh))
  - Configure Dependabot. ([#341](https://github.com/rubygems/gemstash/pull/341), [@olleolleolle](https://github.com/olleolleolle))

## 2.2.2 (2023-03-03)

### Changes

  - Allow newer versions of the `activesupport` dependency ([#332](https://github.com/rubygems/gemstash/pull/332), [@kyrofa](https://github.com/kyrofa))

## 2.2.1 (2023-02-09)

### Bugfixes

  - Add permitted classes to a `YAML.load` in cache ([#329](https://github.com/rubygems/gemstash/pull/329), [@SocalNick](https://github.com/SocialNick))

## 2.2.0 (2023-01-27)

### Bugfixes

  - Respect auth set in upstream host uri ([#317](https://github.com/rubygems/gemstash/pull/317), [@jebentier](https://github.com/jbentier))
  - Change the response code for "/versions" from 403 to 404 ([#326](https://github.com/rubygems/gemstash/pull/326), [@hendric-dev](https://github.com/hendric-dev))

### Changes

  - Support Ruby 3 ([#312](https://github.com/rubygems/gemstash/pull/312), [@tonytonyjan](https://github.com/tonytonyjan))
  - Drop support for EOL Ruby versions 2.4, 2.5, 2.6 ([#328](https://github.com/rubygems/gemstash/pull/328), [@indirect](https://github.com/indirect))
  - Update `dalli` dependency ([#324](https://github.com/rubygems/gemstash/pull/324), [@indirect](https://github.com/indirect))
  - Also thanks to: [@indirect][] and [@hsbt][] who fixed CI issues and lint warnings.

## 2.1.0 (2020-02-26)

### Bugfixes

  - Shebang should be the first thing in the file ([#222](https://github.com/rubygems/gemstash/pull/222), [@bronzdoc](https://github.com/bronzdoc))

### Changes

  - Update rack-test requirement from ~> 0.6 to ~> 1.1 ([#199](https://github.com/rubygems/gemstash/pull/199), [@dependabot](https://github.com/dependabot))
  - Update rake requirement from ~> 10.0 to ~> 12.3 ([#200](https://github.com/rubygems/gemstash/pull/200), [@dependabot](https://github.com/dependabot))
  - CI matrix: Drop 2.0, 2.1, 2.2 ([#201](https://github.com/rubygems/gemstash/pull/201), [@olleolleolle](https://github.com/olleolleolle))
  - Update rubocop version and update files to be inline with new cops ([#202](https://github.com/rubygems/gemstash/pull/202), [@bronzdoc](https://github.com/bronzdoc))
  - Update rubocop requirement from = 0.60.0 to = 0.61.0 ([#208](https://github.com/rubygems/gemstash/pull/208), [@dependabot](https://github.com/dependabot))
  - Update rubocop requirement from = 0.61.0 to = 0.61.1 ([#209](https://github.com/rubygems/gemstash/pull/209), [@dependabot](https://github.com/dependabot))
  - Avoid Rubocop warnings ([#210](https://github.com/rubygems/gemstash/pull/210), [@olleolleolle](https://github.com/olleolleolle))
  - RSpec: config.disable_monkey_patching! ([#211](https://github.com/rubygems/gemstash/pull/211), [@olleolleolle](https://github.com/olleolleolle))
  - Travis: update JRuby to 9.2.5.0 ([#212](https://github.com/rubygems/gemstash/pull/212), [@olleolleolle](https://github.com/olleolleolle))
  - Update rubocop requirement from = 0.61.1 to = 0.62.0 ([#213](https://github.com/rubygems/gemstash/pull/213), [@dependabot](https://github.com/dependabot))
  - Update rubocop requirement from = 0.62.0 to = 0.63.0 ([#215](https://github.com/rubygems/gemstash/pull/215), [@dependabot](https://github.com/dependabot))
  - gemspec: Allow Bundler 2 ([#217](https://github.com/rubygems/gemstash/pull/217), [@olleolleolle](https://github.com/olleolleolle))
  - Update rubocop requirement from = 0.63.0 to = 0.63.1 ([#218](https://github.com/rubygems/gemstash/pull/218), [@dependabot](https://github.com/dependabot))
  - Update rubocop requirement from = 0.63.1 to = 0.64.0 ([#220](https://github.com/rubygems/gemstash/pull/220), [@dependabot](https://github.com/dependabot))
  - CI: matrix updates - update JRuby, add 2.6.1 ([#221](https://github.com/rubygems/gemstash/pull/221), [@olleolleolle](https://github.com/olleolleolle))

lib/gemstash/configuration.rb
# frozen_string_literal: true

require "yaml"
require "erb"

module Gemstash
  # :nodoc:
  class Configuration
    DEFAULTS = {
      cache_type: "memory",
      base_path: File.expand_path("~/.gemstash"),
      db_adapter: "sqlite3",
      bind: "tcp://0.0.0.0:9292",
      rubygems_url: "https://rubygems.org",
      ignore_gemfile_source: false,
      protected_fetch: false,
      fetch_timeout: 20,
      open_timeout: 2,
      # Actual default for db_connection_options is dynamic based on the adapter
      db_connection_options: {},
      puma_threads: 16,
      puma_workers: 1,
      cache_expiration: 30 * 60,
      cache_max_size: 500
    }.freeze

    DEFAULT_FILE = File.expand_path("~/.gemstash/config.yml").freeze

    # This error is thrown when a config file is explicitly specified that
    # doesn't exist.
    class MissingFileError < StandardError
      def initialize(file)
        super("Missing config file: #{file}")
      end
    end

    def initialize(file: nil, config: nil)
      if config
        @config = DEFAULTS.merge(config).freeze
        return
      end

      raise MissingFileError, file if file && !File.exist?(file)

      file ||= default_file

      if File.exist?(file)
        @config = parse_config(file)
        @config = DEFAULTS.merge(@config)
        @config.freeze
      else
        @config = DEFAULTS
      end
    end

    def default?(key)
      @config[key] == DEFAULTS[key]
    end

    def [](key)
      @config[key]
    end

    def keys
      @config.keys
    end

    # @return [Hash] Sequel connection configuration hash
    def database_connection_config
      case self[:db_adapter]
      when "sqlite3"
        { max_connections: 1 }.merge(self[:db_connection_options])
      when "postgres", "mysql", "mysql2"
        { max_connections: (self[:puma_workers] * self[:puma_threads]) + 1 }.merge(self[:db_connection_options])
      else
        raise "Unsupported DB adapter: '#{self[:db_adapter]}'"
      end
    end

  private

    def default_file
      # Support the config file being specified via environment variable
      gemstash_config = ENV["GEMSTASH_CONFIG"]
      return gemstash_config if gemstash_config

      # If no environment variable is used, fall back to the normal defaults
      File.exist?("#{DEFAULT_FILE}.erb") ? "#{DEFAULT_FILE}.erb" : DEFAULT_FILE
    end

    def parse_config(file)
      if file.end_with?(".erb")
        YAML.load(ERB.new(File.read(file)).result) || {}
      else
        YAML.load_file(file) || {}
      end
    end
  end
end
